File: frontend/src/lib/api/rbac.ts
// src/lib/api/rbac.ts — Role-Based Access Control API
import { fetchClient } from './client';

export interface Permission {
  id: string;
  module: string;
  action: string;
  description?: string;
}

export interface RolePermission {
  id: string;
  role_id: string;
  permission_id: string;
  permission: Permission;
}

export interface Role {
  id: string;
  name: string;
  description?: string;
  is_system_default: boolean;
  created_at: string;
  permissions: RolePermission[];
}

export const RBACAPI = {
  getPermissions: () => fetchClient("/rbac/permissions"),
  getRoles: () => fetchClient("/rbac/roles"),
  createRole: (data: { name: string, description?: string, permission_ids: string[] }) => fetchClient("/rbac/roles", { method: "POST", body: JSON.stringify(data) }),
  updateRole: (id: string, data: { name?: string, description?: string, permission_ids?: string[] }) => fetchClient(`/rbac/roles/${id}`, { method: "PUT", body: JSON.stringify(data) }),
  deleteRole: (id: string) => fetchClient(`/rbac/roles/${id}`, { method: "DELETE" }),
};

File: backend/app/models/rbac.py
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy import String, DateTime, Boolean, ForeignKey, Table, Column
from datetime import datetime, timezone
import uuid
from app.models.base import Base

class Role(Base):
    """
    Custom roles created by the Tenant Admin to assign granular access control.
    """
    __tablename__ = "roles"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: uuid.uuid4().hex)
    name: Mapped[str] = mapped_column(String(100), index=True, nullable=False)
    description: Mapped[str] = mapped_column(String(255), nullable=True)
    is_system_default: Mapped[bool] = mapped_column(Boolean, default=False) # e.g. "Employee", "HR Admin" cannot be deleted
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    # Relations
    permissions = relationship("RolePermission", back_populates="role", cascade="all, delete-orphan")

class Permission(Base):
    """
    System-defined permissions that can be granted to roles (e.g., 'users:read', 'it_assets:write').
    """
    __tablename__ = "permissions"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: uuid.uuid4().hex)
    module: Mapped[str] = mapped_column(String(50), index=True, nullable=False) # e.g., 'people', 'finance', 'it'
    action: Mapped[str] = mapped_column(String(50), nullable=False) # e.g., 'create', 'read', 'update', 'delete', 'manage'
    description: Mapped[str] = mapped_column(String(255), nullable=True)

class RolePermission(Base):
    """
    Association table linking Roles to Permissions.
    """
    __tablename__ = "role_permissions"

    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: uuid.uuid4().hex)
    role_id: Mapped[str] = mapped_column(String, ForeignKey("roles.id", ondelete="CASCADE"), index=True, nullable=False)
    permission_id: Mapped[str] = mapped_column(String, ForeignKey("permissions.id", ondelete="CASCADE"), index=True, nullable=False)

    role = relationship("Role", back_populates="permissions")
    permission = relationship("Permission")
